### ISSUE TITLE
Add Zstandard compression support

### ISSUE THREAD

**Issue by svanzoest** [aptly-dev/aptly#989]
As mentioned on [aptly-discuss](https://groups.google.com/g/aptly-discuss/c/OM4thHxZafs/m/UQfZnQZtAAAJ) aptly does not seem to handle debian packages containing `data.tar.zst` files.

## Detailed Description

aptly fails with the following error message
```
[!] Failed to generate package contents:  unsupported tar compression in apt_2.3.9_amd64.deb: data.tar.zst
```
All Ubuntu 21.10 and newer debian packages use [Zstandard](https://facebook.github.io/zstd/) as described in [Hello zstd compressed .debs in Ubuntu!](https://balintreczey.hu/blog/hello-zstd-compressed-debs-in-ubuntu/)

## Context

In order to use aptly to create an Ubuntu mirror after Ubuntu 21.10, support for Zstandard will be needed.

from `deb(5)`
```
       The second required member is named control.tar.  It is a tar archive containing the package control information, either not compressed
       (supported since dpkg 1.17.6), or compressed with gzip (with .gz extension), xz (with .xz extension, supported since  1.17.6)  or  zstd
       (with .zst extension, supported since 1.19.0.5ubuntu2), as a series of plain files, of which the file control is mandatory and contains
       the core control information, the conffiles, triggers, shlibs and symbols files contain optional control information, and the  preinst,
       postinst,  prerm  and  postrm  files are optional maintainer scripts.  The control tarball may optionally contain an entry for ‘.’, the
       current directory.

       The third, last required member is named data.tar.  It contains the filesystem as a tar archive, either not compressed (supported since
       dpkg  1.10.24),  or  compressed  with  gzip (with .gz extension), xz (with .xz extension, supported since dpkg 1.15.6), zstd (with .zst
       extension, supported since 1.19.0.5ubuntu2), bzip2 (with .bz2 extension, supported since dpkg 1.10.24) or lzma (with  .lzma  extension,
       supported since dpkg 1.13.25).
```

## Possible Implementation

- Add support for Zstandard compressed control files by adding to the switch statement in the `GetControlFileFromDeb` function in  [deb.go](https://github.com/aptly-dev/aptly/blob/master/deb/deb.go#L60-L79)
- Add support for Zstandard compressed data files by adding to the switch statement in the `GetContentsFromDeb` function in [deb.go](https://github.com/aptly-dev/aptly/blob/master/deb/deb.go#L165-L194)

This can be done using one of these 3 Go packages as described at https://facebook.github.io/zstd/

| Language | Author | URL
| -- | -- | --
| Go (Port) | Klaus Post | https://github.com/klauspost/compress/tree/master/zstd#zstd
| Go | Vianney Tran | https://github.com/DataDog/zstd
| Go (+dict) | Aliaksandr Valialkin | https://github.com/valyala/gozstd

**Comment by NeroBurner** [comment]
Until the zst support is merged into `aptly` there is a workaround (when building the packages yourself). Workaround found here: https://raphaelhertzog.com/2010/09/17/how-to-create-debian-packages-with-alternative-compression-methods/

add to `debian/rules`:

```make
override_dh_builddeb:
	dh_builddeb -- -Zxz
```

This creates debs compressed with `*.xz`

**Comment by ahaswell** [comment]
I've staged this change in #1002 as the previous pr looks abandoned. I think it's good and just needs merging now.

**Comment by eilandert** [comment]
I can confirm it is working

I've now a functional zstd aptly [packages](https://deb.myguard.nl/pool/main/a/aptly/) (and [docker](https://hub.docker.com/r/eilandert/aptly)), including a new dependency package [golang-github-klauspost-compress](https://deb.myguard.nl/pool/main/g/golang-github-klauspost-compress/)

As soon as this is merged I can switch back to the aptly-dev repo 

Thanks.

**Comment by mbearup** [comment]
Created a fresh PR #1050 since #1002 seems to be stalled. Whatever it takes to get this feature merged :)

**Comment by hcgrove** [comment]
PR #1050 has been merged.

We just need a new release to make this more easily available.

**Comment by TheChef23** [comment]
That's great. Is this available in nightly builds? I could use those until an official release is available.

**Comment by randombenj** [comment]
Nightly publishing should now work, and the latest nightly should have the Zstandard compression support published

**Comment by NeroBurner** [comment]
Tried aptly nightly version `1.4.0+176+g4c04e774` on Ubuntu 22.04 and it works with `zstd`-compressed debian packages now :tada: 

Ubuntu 22.04 jammy provides aptly `1.4.0+ds1-6 amd64` which apt thinks is newer than `1.4.0+176+g4c04e774`.
To prevent apt to update (or rather downgrade) the package, freeze the package using the following command:
```sh
sudo apt-mark hold aptly
```

**Comment by jabl** [comment]
Can confirm the nightly build now works with Ubuntu 22.04 zstd compressed packages. 

Can we have a new release, I imagine this is an issue lots of users are going to run into (and also due to the version mixup with the version published in 22.04 as mentioned by NeroBurner)?

**Comment by randombenj** [comment]
Thanks for all the feedback, as 1.5.0 is released I will close the issue.
